Bottom: f3b4ce6b6c40fbbd6f3650ee0cf1dc64764fbd61
Top:    61b1ad9084149f510114136557d9f407f5073ee3
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-08-20 16:30:18 +0100

hid/gtk: Allow zooming out past the board size


---

diff --git a/src/hid/gtk/gtkhid-main.c b/src/hid/gtk/gtkhid-main.c
index 27ceeb9..7790209 100644
--- a/src/hid/gtk/gtkhid-main.c
+++ b/src/hid/gtk/gtkhid-main.c
@@ -35,26 +35,13 @@ bool ghid_flip_x = false, ghid_flip_y = false;
 static void ghid_zoom_view_fit (void);
 
 static void
-ghid_pan_view_abs (Coord pcb_x, Coord pcb_y, int widget_x, int widget_y)
+pan_common (GHidPort *port)
 {
-  gport->view_x0 = MAX (0, SIDE_X (pcb_x) - widget_x * gport->zoom);
-  gport->view_y0 = MAX (0, SIDE_Y (pcb_y) - widget_y * gport->zoom);
-
-  /* Don't pan so far to the right or bottom that we see past the board edge */
-  gport->view_x0 = MIN (gport->view_x0, PCB->MaxWidth  - gport->view_width);
-  gport->view_y0 = MIN (gport->view_y0, PCB->MaxHeight - gport->view_height);
-
-  /* Don't view above or to the left of the board... ever */
-  gport->view_x0 = MAX (0, gport->view_x0);
-  gport->view_y0 = MAX (0, gport->view_y0);
-
-  /* If we can see the entire board and some, then zoom to fit */
-  if (gport->view_width  > PCB->MaxWidth  &&
-      gport->view_height > PCB->MaxHeight)
-    {
-      ghid_zoom_view_fit ();
-      return;
-    }
+  /* Don't pan so far the board is completely off the screen */
+  port->view_x0 = MAX (-port->view_width,  port->view_x0);
+  port->view_y0 = MAX (-port->view_height, port->view_y0);
+  port->view_x0 = MIN ( port->view_x0, PCB->MaxWidth);
+  port->view_y0 = MIN ( port->view_y0, PCB->MaxHeight);
 
   ghidgui->adjustment_changed_holdoff = TRUE;
   gtk_range_set_value (GTK_RANGE (ghidgui->h_range), gport->view_x0);
@@ -64,6 +51,15 @@ ghid_pan_view_abs (Coord pcb_x, Coord pcb_y, int widget_x, int widget_y)
   ghid_port_ranges_changed();
 }
 
+static void
+ghid_pan_view_abs (Coord pcb_x, Coord pcb_y, int widget_x, int widget_y)
+{
+  gport->view_x0 = SIDE_X (pcb_x) - widget_x * gport->zoom;
+  gport->view_y0 = SIDE_Y (pcb_y) - widget_y * gport->zoom;
+
+  pan_common (gport);
+}
+
 
 /* gport->zoom:
  * zoom value is PCB units per screen pixel.  Larger numbers mean zooming
@@ -72,6 +68,7 @@ ghid_pan_view_abs (Coord pcb_x, Coord pcb_y, int widget_x, int widget_y)
  * gport->view_width and gport->view_height are in PCB coordinates
  */
 
+#define ALLOW_ZOOM_OUT_BY 5
 static void
 ghid_zoom_view_abs (Coord center_x, Coord center_y, double new_zoom)
 {
@@ -84,7 +81,7 @@ ghid_zoom_view_abs (Coord center_x, Coord center_y, double new_zoom)
    */
   min_zoom = 1;
   max_zoom = MAX (PCB->MaxWidth  / gport->width,
-                  PCB->MaxHeight / gport->height);
+                  PCB->MaxHeight / gport->height) * ALLOW_ZOOM_OUT_BY;
   new_zoom = MIN (MAX (min_zoom, new_zoom), max_zoom);
 
   if (gport->zoom == new_zoom)
@@ -97,15 +94,11 @@ ghid_zoom_view_abs (Coord center_x, Coord center_y, double new_zoom)
   pixel_slop = new_zoom;
   ghid_port_ranges_scale (FALSE);
 
-  gport->view_x0 = MAX (0, SIDE_X (center_x) - xtmp * gport->view_width);
-  gport->view_y0 = MAX (0, SIDE_Y (center_y) - ytmp * gport->view_height);
+  gport->view_x0 = SIDE_X (center_x) - xtmp * gport->view_width;
+  gport->view_y0 = SIDE_Y (center_y) - ytmp * gport->view_height;
 
-  ghidgui->adjustment_changed_holdoff = TRUE;
-  gtk_range_set_value (GTK_RANGE (ghidgui->h_range), gport->view_x0);
-  gtk_range_set_value (GTK_RANGE (ghidgui->v_range), gport->view_y0);
-  ghidgui->adjustment_changed_holdoff = FALSE;
+  pan_common (gport);
 
-  ghid_port_ranges_changed ();
   ghid_set_status_line_label ();
 }
 
@@ -1574,6 +1567,9 @@ Center(int argc, char **argv, Coord pcb_x, Coord pcb_y)
   widget_x = gport->width / 2;
   widget_y = gport->height / 2;
 
+  printf ("Pointer_x: %i, Pointer_y: %i\n", widget_x, widget_y);
+  pcb_printf ("PCB coords: %#mD\n", pcb_x, pcb_y);
+
   ghid_pan_view_abs (pcb_x, pcb_y, widget_x, widget_y);
 
   /* Now move the mouse pointer to the place where the board location
diff --git a/src/hid/gtk/gui-output-events.c b/src/hid/gtk/gui-output-events.c
index 527945a..e6cfaa4 100644
--- a/src/hid/gtk/gui-output-events.c
+++ b/src/hid/gtk/gui-output-events.c
@@ -88,6 +88,7 @@ ghid_port_ranges_pan (gdouble x, gdouble y, gboolean relative)
       y1 = y;
     }
 
+#if 0
   if (x1 < h_adj->lower)
     x1 = h_adj->lower;
   if (x1 > h_adj->upper - h_adj->page_size)
@@ -97,6 +98,7 @@ ghid_port_ranges_pan (gdouble x, gdouble y, gboolean relative)
     y1 = v_adj->lower;
   if (y1 > v_adj->upper - v_adj->page_size)
     y1 = v_adj->upper - v_adj->page_size;
+#endif
 
   ghidgui->adjustment_changed_holdoff = TRUE;
   gtk_range_set_value (GTK_RANGE (ghidgui->h_range), x1);
@@ -125,24 +127,28 @@ ghid_port_ranges_scale (gboolean emit_changed)
   gport->view_width = gport->width * gport->zoom;
   gport->view_height = gport->height * gport->zoom;
 
-  if (gport->view_width >= PCB->MaxWidth)
-    gport->view_width = PCB->MaxWidth;
-  if (gport->view_height >= PCB->MaxHeight)
-    gport->view_height = PCB->MaxHeight;
+#if 0
+  if (gport->view_width >= 5 * PCB->MaxWidth)
+    gport->view_width = 5 * PCB->MaxWidth;
+  if (gport->view_height >= 5 * PCB->MaxHeight)
+    gport->view_height = 5 * PCB->MaxHeight;
+#endif
 
   adj = gtk_range_get_adjustment (GTK_RANGE (ghidgui->h_range));
-  adj->page_size = gport->view_width;
-  adj->page_increment = adj->page_size/10.0;
-  adj->step_increment = adj->page_size/100.0;
-  adj->upper = PCB->MaxWidth;
+  adj->page_size = MIN (gport->view_width, PCB->MaxWidth);
+  adj->page_increment = adj->page_size / 10.0;
+  adj->step_increment = adj->page_size / 100.0;
+  adj->lower = -gport->view_width;
+  adj->upper = PCB->MaxWidth + adj->page_size;
   if (emit_changed)
     gtk_signal_emit_by_name (GTK_OBJECT (adj), "changed");
 
   adj = gtk_range_get_adjustment (GTK_RANGE (ghidgui->v_range));
-  adj->page_size = gport->view_height;
-  adj->page_increment = adj->page_size/10.0;
-  adj->step_increment = adj->page_size/100.0;
-  adj->upper = PCB->MaxHeight;
+  adj->page_size = MIN (gport->view_height, PCB->MaxHeight);
+  adj->page_increment = adj->page_size / 10.0;
+  adj->step_increment = adj->page_size / 100.0;
+  adj->lower = -gport->view_height;
+  adj->upper = PCB->MaxHeight + adj->page_size;
   if (emit_changed)
     gtk_signal_emit_by_name (GTK_OBJECT (adj), "changed");
 }
